ios/engine/KMEI/KeymanEngineTests/QueryPackageVersionTests.swift

```diff
@@ -0,0 +1,74 @@
+//
+//  QueryPackageVersionTests.swift
+//  KeymanEngineTests
+//
+//  Created by Joshua Horton on 6/25/20.
+//  Copyright © 2020 SIL International. All rights reserved.
+//
+
+import XCTest
+@testable import KeymanEngine
+
+class QueryPackageVersionTests: XCTestCase {
+  let PACKAGE_HOST_DOMAIN = "https://downloads.keyman.com/"
+  /**
+   * A rigorous test of our package-version query code that runs against the actual, live api.keyman.com endpoint.
+   */
+  func testFetch() throws {
+    let fullIDs = [TestUtils.Keyboards.khmer_angkor.fullID,
+                   TestUtils.Keyboards.sil_euro_latin.fullID,
+                   TestUtils.LexicalModels.mtnt.fullID]
+    let expectation = XCTestExpectation(description: "Package-version query results received")
+    Queries.PackageVersion.fetch(for: fullIDs) { results, error in
+      XCTAssertNil(error)
+      XCTAssertNotNil(results)
+
+      if let results = results {
+        XCTAssertNotNil(results.keyboards)
+        XCTAssertNotNil(results.models)
+
+        if let keyboardVersions = results.keyboards, let modelVersions = results.models {
+          XCTAssertEqual(keyboardVersions.count, 2)
+          XCTAssertEqual(modelVersions.count, 1)
+
+          keyboardVersions.forEach { entry in
+            if let error = entry.value as? Queries.PackageVersion.ResultError {
+              XCTFail(String(describing: error))
+            }
+          }
+
+          if let khmer_angkor = keyboardVersions["khmer_angkor"] as? Queries.PackageVersion.ResultEntry {
+            XCTAssertGreaterThanOrEqual(Version(khmer_angkor.version)!,
+                                        Version(TestUtils.Keyboards.khmer_angkor.version)!)
+            XCTAssertTrue(khmer_angkor.packageURL.hasPrefix(self.PACKAGE_HOST_DOMAIN))
```

Comment:
This code really shouldn't care what the domain is: if we want to change it in the future from the API side, we shouldn't need to be updating the code. That is, this test could fit in an API test on api.keyman.com, but doesn't belong in client test code.

Reply:
> 
> 
> > but also in the short-term for finding keyboard KMPs
> 
> I'm curious as to how you envisage using the API for this purpose?

The current in-app iOS keyboard search downloads resources based on the selected keyboard's ID.  Unfortunately, the keyboard cloud API returns keyboard IDs, but not package URLs.   That's where this API comes in - it can find the package URL for a given keyboard ID, bridging the gap.

So, as a stop-gap until the keyboard search is replaced:  

1.  The old repository query retrieves data for all languages.  (Fetch 1)
2.  The user then selects a keyboard to install.
3.  (Stopgap) The engine then queries this endpoint for the most current version of that keyboard's package, based on the selected keyboard's ID.  (Fetch 2)
4.  Having retrieved the filepath for the keyboard's package, it is then queued for actual download (Fetch 3)

If we want to get detailed...

5.  Once the keyboard is downloaded, the engine then queries for matching lexical models (Fetch 4)
6.  If one exists, we download and install that resource, too.  (Fetch 5)

Note that all but "fetch 2" are already happening.  Once we convert to using the web-based keyboard search, 1 and 2 will be dropped, as the WebView will automatically forward the data needed for "fetch 3" to proceed.  Fetches 4 and 5 (for lexical models) will be maintained.